72_config_and_simplify_tls.md

```diff
@@ -0,0 +1,358 @@
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [keylime/enhancements] referencing this enhancement and targeting a release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Keylime i.e., [keylime/keylime], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to pull request in
+  [keylime/enhancements]
+- [ ] Core members have approved the issue with the label `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+The `keylime.conf` is migrated to TOML and split into the different components.
+The new configuration format is versioned to allow for simpler upgrades in the
+future.
+
+The current TLS setup uses one CA for the verifier, one for the registrar and
+one for the agent mTLS authentication. This will be replaced by a client and
+server (key and certificate) option in each component (agent, verifier,
+registrar, tenant). The client pair is used to authenticate the component using
+mTLS and the server pair is used for TLS server (like the registrar, agent and
+verifier have). Also two new options are introduced, one for trustworthy
+incoming CAs for server side mTLS authentication and another one for trustworthy
+CAs from the client side.
+
+## Motivation
+
+Keylime uses the Python specific INI inspired configuration format and mixes
+different data types (comma separated list vs. Python syntax). This makes it
+difficult to share the configuration between the Python and Rust code.
+Furthermore the configuration for different components are not separated and
+there is no nice way to upgrade users to newly introduced configuration options.
+
+The current way TLS is setup in Keylime is confusing to newcomers. Also the
+different configuration options are not consistent in their naming scheme.
+
+
+### Goals
+
+* Move the `keylime.conf` to TOML
+* Split the config into five new configuration files: global, agent, tenant,
+  registrar, verifier
```

Comment:
IMHO should not be different files, but add the option of adding snippets. This is more powerful.

For example, we can deploy a "safe" enough config file for the agent service, with the sections for tenant, registrar and verifier empty, but with the value for address of the control plane with an invalid address (like SERVER_ADDRESS).

In that case, the admin only need to do something like

```bash
echo '[agent]' > /etc/keylime.d/revocation.conf
echo "receive_revocation_ip = 10.0.110.15" >> /etc/keylime.d/revocation.conf
```

Setting other attributes is a mater of creating new snippets.

Reply:
I like the idea of the `keylime.d` directory, but still want to split the files, because it forces us to separate the config more clearly.

Reply:
> but still want to split the files

Note that you can have for free this today. You can have "agent.conf", "tenant.conf", etc files. You only need to extend `CONFIG_FILES` list with the contents of `/usr/etc/keylime/*.conf`

What is more, you can have all the set of files in `/usr/etc/keylime` and the overwrites in `/etc/keylime.d` at the same cost.

Reply:
You are right, we get this for free today. 
I think the proposed idea of putting the tenant, verifier, registrar and agent configs in  `/etc/keylime/` and adding the option to add overrides in `/etc/keylime.d` is a good idea.

Reply:
> putting the tenant, verifier, registrar and agent configs in /etc/keylime/ and adding the option to add overrides in /etc/keylime.d

Right, for completion, one alternative could be:

* `/usr/etc/keylime/*.conf` from where the packager will deliver safe defaults. There we can have a subset of: `["keylime.conf", "agent.conf", "verifier.conf", "tenant.conf", "registrar.conf"]`. It is a subset because for the keylime_agent deployment maybe we only need to have `["keylime.conf", "agent.conf"]`
* `/etc/keylime/*.conf`, same as before, this is for the case that the system do not support `/usr/etc` (Debian?) and/or the user decides to maintain a full set of configurations.
* `/etc/keylime/keylime.d/*.conf`, for the overrides.

This simplify the logic in `config.py`, as we need to search in three places only, and merge them with the priority (form low to high) as enumerated in the list.